src/dart/lib/decoder.dart

```diff
@@ -0,0 +1,894 @@
+// Copyright 2021 The IconVG Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:typed_data';
+import 'dart:ui';
+
+import 'package:flutter/foundation.dart';
+import 'package:vector_math/vector_math_64.dart';
+
+export 'dart:typed_data' show Uint8List;
+export 'dart:ui' show Color, Picture, Rect;
+
+@immutable
+/// IconVG parser.
+///
+/// The constructor takes a [Uint8List] of raw bytes to parse.
+///
+/// The output is a [picture] and the graphic coordinate space [rect].
+///
+/// A custom palette can be provided to the decoder using the `palette` argument
+/// to the constructor. It must not be longer than 64 colors.
+///
+/// If the size that the image will be rendered at is known ahead of time, the
+/// height that will be used can be provided using the `height` argument. The
+/// default is the graphic coordinate space height.
+///
+/// By default the rendered image is clipped without anti-aliasing. This can be
+/// controled using the constructor's `clipBehavior` argument.
+class IconVGFile {
+  const IconVGFile._(this.picture, this.rect);
+
+  factory IconVGFile(Uint8List bytes, {
+    List<Color>? palette,
+    double height = double.infinity,
+    Clip clipBehavior = Clip.hardEdge,
+  }) {
+    assert(palette == null || palette.length <= _IconVGMachine.kCREGLength);
+    final _IconVGMachine machine = _IconVGMachine(bytes, palette: palette, height: height, clipBehavior: clipBehavior);
+    return IconVGFile._(machine.picture, machine.viewBox);
+  }
+
+  /// The parsed image, rendered as a vector graphic image.
+  ///
+  /// This includes a clip as specified by the `clipBehavior` argument to the constructor.
+  final Picture picture;
+
+  /// The view box (graphic coordinate space).
+  ///
+  /// It is common for images to use negative coordinates, so the top left of
+  /// this rectangle may be to the left of and above the origin.
+  final Rect rect;
+
+  static void tests() {
```

Comment:
Consider marking this `@visibleForTesting` and documenting where it's used. Alternatively, consider making the implementation you're testing public or `@visibleForTesting` so that we don't have test code living in the same file as the impl.

Reply:
`@visibleForTesting` is just another way of writing `public`. :-)

I figured this will all get tree-shaken so it doesn't really matter. Added docs though.